File: stack/LinkedStack.py
# Implemntation of Stack using LinkedList

# Node Class to store data and Linked Node
from numpy import true_divide


class Node:
    def __init__(self, data: int) -> None:
        self.data = data
        self.prev = None

# LinkedStack class to maintain Stack


class LinkedStack:
    def __init__(self, size: int) -> None:
        self.size = size
        self.top = None
        self.count = 0

    def push(self, data: int) -> bool:
        if self.count == self.size:
            print("Stack Overflow!")
            return False
        if self.top == None:
            self.top = Node(data)
            self.prev = None
            self.count += 1
            return True
        temp = self.top
        self.top = Node(data)
        self.top.prev = temp
        self.count += 1
        return True

    def pop(self) -> int:
        if self.top == None:
            print("Stack Underflow!")
            return -999
        data = self.top.data
        self.top = self.top.prev
        self.count -= 1
        return data

    def show(self):
        if not self.top:
            print('Stack Empty!')
            return
        temp = self.top
        while temp.prev:
            print(temp.data, end='=>')
            temp = temp.prev
        print(temp.data)

    def peek(self) -> int:
        if self.top:
            return self.top.data
        return -999

    def isEmpty(self) -> bool:
        if self.top:
            return False
        return True

    def free(self):
        return self.size - self.count


if __name__ == '__main__':
    stack = LinkedStack(5)

    while True:
        option = input(
            "\nChoose: 1.Push() 2.Pop() 3.Peek() 4.isEmpty() 5.free() 6.Show() -1. Exit: ")
        if option == '1':
            data = int(input("Data: "))
            stack.push(data)
        elif option == '2':
            print(f'Poped Data: {stack.pop()}')
        elif option == '3':
            print(f'Peek Data: {stack.peek()}')
        elif option == '4':
            print(f'Is Empty: {stack.isEmpty()}')
        elif option == '5':
            print(f'AVailable Space: {stack.free()}')
        elif option == '6':
            stack.show()
        else:
            print("Choose Valid Option")

File: stack/LinkedSTack.cpp
#include <iostream>

class Node
{
public:
    int data;
    Node *prev;
    Node(int data)
    {
        this->data = data;
        this->prev = NULL;
    }
};

// Stack Class
class Stack
{
private:
    int size, count;
    Node *top;

public:
    Stack(int size)
    {
        this->size = size;
        this->count = 0;
        std::cout << "Stack Initialized";
    }

    bool push(int data)
    {
        if (this->count < this->size)
        {
            Node *node = new Node(data);
            if (this->count == 0)
            {

                this->top = node;
                this->top->prev = NULL;
                this->count++;
                return true;
            }
            Node *temp = this->top;
            this->top = node;
            this->top->prev = temp;
            this->count++;
            return true;
        }
        std::cout << "Stack Overflow Error!" << std::endl;
        return false;
    }

    int pop()
    {
        int data;
        if (this->count == 0)
        {
            std::cout << "\nStack Underflow";
            return -999;
        }
        data = this->top->data;
        this->top = this->top->prev;
        this->count--;
        return data;
    }
    int peek()
    {
        if (this->count == 0)
        {
            std::cout << "\nStack Underflow";
            return -999;
        }
        return this->top->data;
    }

    void show()
    {
        if (this->count == 0)
        {
            return;
        }
        Node *temp = this->top;
        while (temp->prev)
        {
            std::cout << " ->" << temp->data;
            temp = temp->prev;
        }
        std::cout << "->" << temp->data << std::endl;
    }

    bool isEmpty()
    {
        if (this->count == 0)
        {
            return true;
        }
        return false;
    }

    int free()
    {
        return this->size - this->count;
    }
};

// main
int main(int argc, char *argv[])
{
    int option, data;
    Stack stack = Stack(5);
    do
    {
        std::cout << "\n=======Choose======" << std::endl;
        std::cout << "1.Push || 2.Pop || 3.show || 4.Peek || 5. isEmpy || 6.free : ";
        std::cin >> option;
        switch (option)
        {
        case 1:
            std::cout << "\n Enter Data:";
            std::cin >> data;
            stack.push(data);
            break;
        case 2:
            data = stack.pop();
            std::cout << "\n Popped: " << data;
            break;
        case 3:
            stack.show();
            break;
        case 4:
            std::cout << "\nData: " << stack.peek();
            break;
        case 5:
            std::cout << "\nEmpty: " << stack.isEmpty();
            break;
        case 6:
            std::cout << "\nAvailable Space: " << stack.free();
            break;
        default:
            std::cout << "\nChoose Valid Option";
            break;
        }

    } while (option != 0);
    return 0;
}
